### ISSUE TITLE
Random divergence of memory usage in minimize_mem=True

### ISSUE THREAD

**Issue by kalaspa** [gplepage/vegas#15]
Hi,
Sorry for posting a second issue on this wonderful library, but here I am with my pathological integrand.
I am performing a 4 dimensional integral with a lot of evaluation, and I have to minimize memory usage. Hence I activate the `minimize_mem=True` option when initializing the integrator. However, I still obtain the following error:
```bash
Traceback (most recent call last):
  File "/home/{hidden}.py", line 606, in <module>
    gw_spectrum(
  File "/home/{hidden}.py", line 534, in gw_spectrum
    integ(integrand, nitn=10, neval=neval)
  File "_vegas.pyx", line 1546, in vegas._vegas.Integrator.__call__
  File "_vegas.pyx", line 1339, in random_batch
numpy.core._exceptions.MemoryError: Unable to allocate 215. GiB for an array with shape (7203260000, 4) and data type float64
```
I was able to reproduce the error with the following minimal version of my code and trace memory usage
```python
import numpy as np
import tracemalloc
import sys
import vegas

@vegas.batchintegrand
def integrand(vegas_vector):
    v_1, v_2, v_3, v_4 = (
        vegas_vector[:, 0],
        vegas_vector[:, 1],
        vegas_vector[:, 2],
        vegas_vector[:, 3],
    )

    return np.cos(v_1) * np.exp(-(v_2 ** 2)) * v_3 ** 3 / (1 + v_3) ** 5


if __name__ == "__main__":

    tracemalloc.start()

    integ = vegas.Integrator(
        [
            [1, 2],
            [0, 10000],
            [0, 10000],
            [0, 10000],
        ],
        mpi=False,
        minimize_mem=True,
    )

    # First train the integrator and throw away the results
    integ(integrand, nitn=10, neval=100)
    # Second run the trained integrator
    integ(integrand, nitn=10, neval=100)

    current, peak = tracemalloc.get_traced_memory()
    print(
        f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB"
    )
    tracemalloc.stop()
```
Most of the time, the memory usage is constant, but once in a while (roughly one out of 10), the memory usage diverges. Here is my output
```bash
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 32.186484MB; Peak was 57.916624MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.279412MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.280292MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.276564MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.276596MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.276708MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.278452MB; Peak was 3.026508MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.28106MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 2.310756MB; Peak was 4.135084MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.2953MB; Peak was 0.70384MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 25.03898MB; Peak was 45.051368MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.276564MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.278452MB; Peak was 79.4753MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.280324MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.278404MB; Peak was 0.605492MB
[username@machine Bureau]$ python minimal_example.py
Current memory usage is 0.280372MB; Peak was 317.985612MB
```
It even crashed my computer once or twice during these tests. This happened on at least two different Linux machines. Can you reproduce this error ? And do you know any workaround ?

**Comment by gplepage** [comment]
I think I know what is going on with the minimal version. The integrand is extremely poorly behaved for ``neval=100`` and so you get consecutive iterations where one estimate is -1.1(1.1)e-120 and the next is 0.290(14). The jump by 119 orders of magnitude defeats the algorithm for saving memory -- it assumes a greater degree of stability from one iteration to the next. I will look to put in protection against memory overflows, but, no matter what I do, ``minimize_mem=True`` won't work very well, if at all, when there is this much instability from iteration to iteration. 

You can limit the large memory allocation by setting ``max_neval_hcube = 1000`` or some other not too large number. This limits the number of function evaluations in a single hypercube, which could limit the algorithm's ability to adapt but prevents it from going crazy in response to a huge fluctuation. The default for this parameter is 1e7. I might lower the default as a result of this discussion. In any case you might try resetting this parameter to a lower number for your real problem and see if that helps.

**Comment by gplepage** [comment]
Version 3.5.4 reduces the default for ``max_neval_hube`` which should make this issue much less dramatic. I also added a mechanism for limiting the internal memory used (``max_mem``) to avoid crashes.

The fluctuations you were seeing with the minimal problem are because ``neval`` is much too small for that integrand. Part of the problem is the integrand vanishes exactly when ``v2`` is larger than 3 or so, while the integration region goes out to 10,000. This means that ``vegas`` can easily miss the integrand completely if ``neval`` is too small. This sort of thing can make the algorithm fluctuate wildly from iteration to iteration. Using ``minimize_mem=True`` makes the situation worse because of the way it works. The problems for this integrand mostly go away if ``neval=1e4``, at least after the first iteration or two (once ``vegas`` sees the peak).

**Comment by gplepage** [comment]
I am closing the issue because it is addressed in v3.5.4 and can be avoided in any case by resetting the default for ``max_neval_hcube``.
